syntax = "proto3";

package NYql.NFmr.NProto;

enum EOperationStatus {
    OPERATION_UNKNOWN = 0;
    OPERATION_ACCEPTED = 1;
    OPERATION_IN_PROGRESS = 2;
    OPERATION_FAILED = 3;
    OPERATION_COMPLETED = 4;
    OPERATION_ABORTED = 5;
    OPERATION_NOT_FOUND = 6;
}

enum ETaskStatus {
    TASK_UNKNOWN = 0;
    TASK_ACCEPTED = 1;
    TASK_IN_PROGRESS = 2;
    TASK_FAILED = 3;
    TASK_COMPLETED = 4;
    TASK_ABORTED = 5;
}

enum ETaskType {
    TASK_TYPE_UNKNOWN = 0;
    TASK_TYPE_DOWNLOAD = 1;
    TASK_TYPE_UPLOAD = 2;
    TASK_TYPE_MERGE = 3;
}

enum EFmrComponent {
    COMPONENT_UNKNOWN = 0;
    COMPONENT_COORDINATOR = 1;
    COMPONENT_WORKER = 2;
    COMPONENT_JOB = 3;
}

enum EFmrReason {
    REASON_UNKNOWN = 0;
    USER_ERROR = 1;
}

message TFmrError {
    EFmrComponent Component = 1;
    EFmrReason Reason = 2;
    string ErrorMessage = 3;
    optional uint32 WorkerId = 4;
    optional string TaskId = 5;
    optional string OperationId = 6;
}

message TYtTableRef {
    string Path = 1;
    string Cluster = 2;
    optional string FilePath = 3;
}

message TFmrTableId {
    string Id = 1;
}

message TFmrTableRef {
    TFmrTableId FmrTableId = 1;
}

message TTableRange {
    string PartId = 1;
    uint64 MinChunk = 2;
    uint64 MaxChunk = 3;
}

message TFmrTableInputRef {
    string TableId = 1;
    repeated TTableRange TableRanges = 2;
}

message TFmrTableOutputRef {
    string TableId = 1;
    string PartId = 2;
}

message TTableStats {
    uint64 Chunks = 1;
    uint64 Rows = 2;
    uint64 DataWeight = 3;
}

message TFmrStatisticsObject {
    TFmrTableOutputRef Table = 1;
    TTableStats Statistic = 2;
}

message TStatistics {
    repeated TFmrStatisticsObject OutputTables = 1;
}

message TOperationTableRef {
    oneof OperationTableRef {
        TYtTableRef YtTableRef = 1;
        TFmrTableRef FmrTableRef = 2;
    }
}

message TTaskTableRef {
    oneof TaskTableRef {
        TYtTableRef YtTableRef = 1;
        TFmrTableInputRef FmrTableInputRef = 2;
    }
}

message TUploadOperationParams {
    TFmrTableRef Input = 1;
    TYtTableRef Output = 2;
}

message TUploadTaskParams {
    TFmrTableInputRef Input = 1;
    TYtTableRef Output = 2;
}

message TDownloadOperationParams {
    TYtTableRef Input = 1;
    TFmrTableRef Output = 2;
}

message TDownloadTaskParams {
    TYtTableRef Input = 1;
    TFmrTableOutputRef Output = 2;
}

message TMergeOperationParams {
    repeated TOperationTableRef Input = 1;
    TFmrTableRef Output = 2;
}

message TMergeTaskParams {
    repeated TTaskTableRef Input = 1;
    TFmrTableOutputRef Output = 2;
}

message TOperationParams {
    oneof TOperationParams {
        TUploadOperationParams UploadOperationParams = 1;
        TDownloadOperationParams DownloadOperationParams = 2;
        TMergeOperationParams MergeOperationParams = 3;
    }
}

message TTaskParams {
    oneof TTaskParams {
        TUploadTaskParams UploadTaskParams = 1;
        TDownloadTaskParams DownloadTaskParams = 2;
        TMergeTaskParams MergeTaskParams = 3;
    }
}

message TClusterConnection {
    string TransactionId = 1;
    string YtServerName = 2;
    optional string Token = 3;
}

message TTask {
    ETaskType TaskType = 1;
    string TaskId = 2;
    TTaskParams TaskParams = 3;
    string SessionId = 4;
    optional uint32 NumRetries = 5;
    map<string, TClusterConnection> ClusterConnections = 6;
    optional string JobSettings = 7;
}

message TTaskState {
    ETaskStatus TaskStatus = 1;
    string TaskId = 2;
    optional TFmrError TaskErrorMessage = 3;
    TStatistics Stats = 4;
}
